yjit/src/codegen.rs

```diff
@@ -9922,6 +9922,34 @@ fn gen_objtostring(
         );
 
         // No work needed. The string value is already on the top of the stack.
+        Some(KeepCompiling)
+    } else if unsafe { RB_TYPE_P(comptime_recv, RUBY_T_SYMBOL) } && assume_method_basic_definition(jit, asm, comptime_recv.class_of(), ID!(to_s)) {
+        jit_guard_known_klass(
+            jit,
+            asm,
+            comptime_recv.class_of(),
+            recv,
+            recv.into(),
+            comptime_recv,
+            SEND_MAX_DEPTH,
+            Counter::objtostring_not_string,
```

Comment:
Small detail: here I think we would want a new counter `objtostring_not_symbol` since we are guarding on the receiver being a symbol.

Reply:
Both names are incorrect, I think. The counter gets passed to jit_chain_guard() to terminate the chain, so the it gets incremented when the max depth is reached. The jump to the counted exit is not conditional, too, so it doesn't actually check anything at runtime to know if it's a string or not.


```
    if asm.ctx.get_chain_depth() < depth_limit {
        ...
    } else {
        target0_gen_fn.call(asm, Target::side_exit(counter), None);
    }
```

There are other instances of this misuse and we should clean them up IMO. If you agree we should leave this as is and do the clean-up serpately.